File: Game.py
from __future__ import annotations

from random import choice, randint, shuffle
from time import sleep
from typing import Any, NamedTuple

from termcolor import colored

from Card import Card, new_deck
from Player import (BeliefPlayer, CooperativePlayer, GreedyPlayer, MaximizePointsPlayer,
                    MaximizeRoundsWonPlayer, Player, PredictorPlayer, RandomPlayer)
from Team import Team

# Command line strategy name -> the class that implements it.
# Adding a strategy means adding its class here and nothing else.
STRATEGIES = {
    'random': RandomPlayer,
    'greedy': GreedyPlayer,
    'maxpointswon': MaximizePointsPlayer,
    'maxroundswon': MaximizeRoundsWonPlayer,
    'cooperative': CooperativePlayer,
    'predictor': PredictorPlayer,
}

# Player ids are 1 and 2 for the first team and 3 and 4 for the second, and are used to
# index the belief arrays, so the order of this table matters.
TEAMS = (('Sporting', ('Leitao', 'Fred')), ('Benfica', ('Pedro', 'Sebas')))

# The player a person controls in human mode
HUMAN_PLAYER = 'Leitao'

CARDS_PER_HAND = 10
ROUNDS_PER_GAME = 10


class RoundResult(NamedTuple):
    '''
        points: total value of the cards on the table
        card: the card winning the round so far
        winner: index of that card in the list of cards played
    '''

    points: int
    card: Card
    winner: int


class Game:
    '''
        Game ->
            - teams: list of Team objects
            - playersOrder: list of Player objects sorted by order to play
            - deck: list of Card objects
            - trump: trump card for that game
            - current_round: index of the round being played (0 - 9)
            - game_info: dictionary with game information
            - verbose: boolean to print game details
            - mode: string with the mode of the game (auto or human)
    '''

    def __init__(self, team_1_strategy: str, team_2_strategy: str, v: bool, mode: str) -> None:
        self.verbose = v
        self.mode = mode

        self.trump = None
        self.current_round = 0

        # Initialize game information
        self.game_info = {'Teams': [], 'Rounds': {}}

        # Create teams and their players
        self.teams = [self.create_team(name, names, strategy)
                      for (name, names), strategy in zip(TEAMS, (team_1_strategy, team_2_strategy))]

        # Randomize players and team to start
        for team in self.teams:
            shuffle(team.players)
        first_team = choice(self.teams)
        second_team = self.teams[1] if first_team is self.teams[0] else self.teams[0]

        # Order players, alternating between teams
        self.playersOrder = [player for pair in zip(first_team.players, second_team.players) for player in pair]

        # Create deck
        self.deck = self.create_deck()

    def create_team(self, team_name: str, player_names: tuple[str, ...], strategy: str) -> Team:
        '''
            Create a team whose players all follow the given strategy
        '''

        if strategy not in STRATEGIES:
            raise ValueError(f'Invalid strategy: {strategy}')

        player_class = STRATEGIES[strategy]
        team = Team(team_name)

        # Ids run 1, 2 for the first team and 3, 4 for the second
        first_id = 1 + 2 * [name for name, _ in TEAMS].index(team_name)
        for offset, name in enumerate(player_names):
            player = player_class(first_id + offset, name, team, self.verbose)
            player.is_human = self.mode == 'human' and name == HUMAN_PLAYER
            team.add_player(player)

        return team

    def get_partner(self, player_name: str) -> Player | None:
        '''
            Get the partner of the player with the given name
        '''

        for team in self.teams:
            for player in team.players:
                if player.name == player_name:
                    return team.get_partner(player)

        return None

    def create_deck(self) -> list[Card]:
        '''
            Create a deck of 40 cards (see Card.py for the ranks, orders and values)
        '''

        return new_deck()

    def rotate_order_to_winner(self, playersOrderList: list[Player], winner: Player) -> list[Player]:
        '''
            Rotate the list of players to the winner of the round
        '''

        winner_index = playersOrderList.index(winner)

        return playersOrderList[winner_index:] + playersOrderList[:winner_index]

    def evaluate_round(self, cardsPlayedInRound: list[Card]) -> RoundResult:
        '''
            Calculate the points on the table and which of the cards is winning them.

            Also used by the strategies to score rounds that are still incomplete, and
            hypothetical ones.
        '''

        winner = 0
        for i, card in enumerate(cardsPlayedInRound[1:], start=1):
            if card.beats(cardsPlayedInRound[winner], self.trump.suit):
                winner = i

        points = sum(card.value for card in cardsPlayedInRound)

        return RoundResult(points, cardsPlayedInRound[winner], winner)

    def hand_cards(self) -> None:
        '''
            Distribute the cards between the players. The last card dealt is the trump
        '''

        card = None
        for player in self.playersOrder:
            for _ in range(CARDS_PER_HAND):
                # Pop a card at random
                card = self.deck.pop(randint(0, len(self.deck) - 1))
                player.add_card(card)
                player.team.initial_points += card.value

                # Update beliefs of the player
                if isinstance(player, BeliefPlayer):
                    player.update_beliefs_initial(card)

        self.trump = card

        # Print game details
        if self.verbose and self.mode == 'auto':
            for player in self.playersOrder:
                print(colored(f'{player.name} -> {player.team.name}', 'yellow', attrs=['underline']))
                print(colored(player.get_strategy(), attrs=['bold']))
                for card in player.hand:
                    print(card.name)
                print('\n')

        if self.verbose or self.mode == 'human':
            print(colored(f'Trump card: {self.trump.name}\n', 'blue', attrs=['bold']))

    def update_beliefs(self, cardPlayed: Card, round_suit: str, player: Player) -> None:
        '''
            Update the beliefs of the players except the one that played the card (no need!)
        '''

        for other in self.playersOrder:
            if other is not player and isinstance(other, BeliefPlayer):
                other.update_beliefs(cardPlayed, round_suit, player, self.mode)

    def ask_human_card(self, player: Player, position: int, round_suit: str, suggestion: Card) -> Card:
        '''
            Show the engine's suggestion and let the person pick a legal card instead
        '''

        # The suggestion was only a suggestion: put it back before asking
        player.add_card(suggestion)

        # Following suit is mandatory whenever the player can
        legal_cards = player.get_cards_by_suit(round_suit) if position > 0 else []
        if not legal_cards:
            legal_cards = list(player.hand)

        print(colored('Your current hand:', 'yellow'))
        for card in player.hand:
            print(colored(card.name, attrs=['bold']))
        print('\nThe engine suggests you play', end=' ')
        print(colored(suggestion.name, 'blue', attrs=['bold']))

        while True:
            card = player.get_card(input(colored('> ', attrs=['bold'])))
            if card is None:
                print(colored('Invalid card! Try again', 'red'))
            elif card not in legal_cards:
                print(colored(f'You have to follow {round_suit}! Try again', 'red'))
            else:
                break

        player.hand.remove(card)
        print(colored(f'You played {card.name}', 'green', attrs=['bold']))

        return card

    def play_round(self, num_round: int) -> dict[str, Any]:
        '''
            Play a round of the game
        '''

        self.current_round = num_round
        roundSuit = ''
        cardsPlayedInRound = []

        for position, player in enumerate(self.playersOrder):
            card_played, roundSuit = player.play_round(position, roundSuit, cardsPlayedInRound, self)

            # In human mode the engine only advises, the person decides
            if player.is_human:
                card_played = self.ask_human_card(player, position, roundSuit, card_played)
                if position == 0:
                    roundSuit = card_played.suit

            cardsPlayedInRound.append(card_played)

            # Update the beliefs of the players
            self.update_beliefs(card_played, roundSuit, player)

            # Give the person time to follow what happened
            if self.mode == 'human':
                sleep(2)

        # Get the total points played in the round and the respective winner
        result = self.evaluate_round(cardsPlayedInRound)
        winner = self.playersOrder[result.winner]
        winner.team.score += result.points

        if self.verbose or self.mode == 'human':
            print(colored('You win the round' if winner.is_human else f'{winner.name} wins the round',
                          'blue', attrs=['bold']))

        # Rotate the players order to the winner of the round
        self.playersOrder = self.rotate_order_to_winner(self.playersOrder, winner)

        return {'Winner': winner.name, 'Points': result.points}

    def play_game(self) -> str:
        '''
            Play the game of Sueca and return the name of the winning team, or "ties"
        '''

        for num_round in range(ROUNDS_PER_GAME):
            if self.verbose or self.mode == 'human':
                print(colored(f'\nRound {num_round + 1}:', 'green', attrs=['underline']))

            self.game_info['Rounds'][num_round + 1] = self.play_round(num_round)

        sporting, benfica = self.teams
        self.game_info['Teams'] = [sporting.dump_to_json(), benfica.dump_to_json()]

        if self.verbose or self.mode == 'human':
            print(colored(f'\n{sporting.name} score: {sporting.score}', 'green'))
            print(colored(f'{benfica.name} score: {benfica.score}', 'red'))
            print()

            if sporting.score > benfica.score:
                print(colored(f'{sporting.name} wins', 'white', 'on_green'))
            elif sporting.score < benfica.score:
                print(colored(f'{benfica.name} wins!', 'white', 'on_red'))
            else:
                print(colored("It's a tie!", 'white', 'on_dark_grey'))

        if sporting.score > benfica.score:
            return sporting.name
        if sporting.score < benfica.score:
            return benfica.name

        return 'ties'

File: android/engine/src/main/kotlin/pt/up/fe/asma/sueca/engine/Game.kt
package pt.up.fe.asma.sueca.engine

import kotlin.random.Random

/**
 * Knobs that separate "reproduce the simulator exactly" from "give good advice at a real table".
 *
 * @param selfBeliefPolicy see [SelfBeliefPolicy].
 * @param perfectInformation when true the Deck Predictor reads the other hands, as the Python
 *   version does; when false it reasons only from its beliefs.
 * @param earlyTrumpAvoidance the `current_round < 2` trump penalty of `PredictorPlayer`.
 * @param revealTrump whether every seat is told who holds the trump card. True at a real table,
 *   where the trump is turned face up, false in the simulator, which never models it.
 * @param maxBranch cap on how many answers per opponent the search expands.
 * @param searchBudget cap on the total number of answers expanded per decision.
 */
data class EngineConfig(
    val selfBeliefPolicy: SelfBeliefPolicy = SelfBeliefPolicy.LEGACY,
    val perfectInformation: Boolean = true,
    val earlyTrumpAvoidance: Boolean = true,
    val revealTrump: Boolean = false,
    val maxBranch: Int = 12,
    val searchBudget: Long = 20_000L,
) {

    val oracle: HandOracle get() = if (perfectInformation) PerfectInfoOracle else BeliefOracle

    companion object {
        /** Bit for bit the behaviour of `sueca.py`, for reproducing the report's numbers. */
        val SIMULATION = EngineConfig()

        /** Agents that do not peek at anyone's hand. Used when a person is at the table. */
        val FAIR = EngineConfig(perfectInformation = false, selfBeliefPolicy = SelfBeliefPolicy.TRACK_OWN_PLAYS)

        /** Tracking a real game: nothing is known beyond what has been seen. */
        val ADVISOR = EngineConfig(
            selfBeliefPolicy = SelfBeliefPolicy.TRACK_OWN_PLAYS,
            perfectInformation = false,
            revealTrump = true,
        )
    }
}

/**
 * @param points total value of the cards on the table.
 * @param card the card winning the round so far.
 * @param winner index of that card in the list of cards played.
 */
data class RoundResult(val points: Int, val card: Card, val winner: Int)

data class Play(val player: Player, val card: Card)

data class Trick(val number: Int, val plays: List<Play>, val winner: Player, val points: Int) {
    val cards: List<Card> get() = plays.map { it.card }
}

data class GameResult(val winner: TeamId?, val scores: Map<TeamId, Int>, val dealt: Map<TeamId, Int>) {
    val isTie: Boolean get() = winner == null
}

/**
 * A game of Sueca, driven one card at a time.
 *
 * Where `Game.py` runs the whole thing in a blocking loop, this exposes the same rules as a
 * state machine ([currentPlayer], [legalCards], [play]) so a UI can wait for a tap, animate a
 * trick, or let an agent take over any seat at any moment. [playToCompletion] recovers the
 * original loop for simulations.
 */
class Game(
    val sportingAgent: AgentKind,
    val benficaAgent: AgentKind,
    val config: EngineConfig = EngineConfig.SIMULATION,
    val random: Random = Random.Default,
    humanPlayerName: String? = null,
) {

    val teams: List<Team> = TeamId.entries.map { id ->
        Team(id, if (id == TeamId.SPORTING) sportingAgent else benficaAgent)
    }

    val sporting: Team get() = teams[TeamId.SPORTING.ordinal]
    val benfica: Team get() = teams[TeamId.BENFICA.ordinal]

    /** All four seats, in id order (1..4), independent of who is playing first. */
    val players: List<Player>

    /** The four seats sorted by who plays next; rotated to the winner after every trick. */
    var playersOrder: List<Player> = emptyList()
        private set

    private val deck: MutableList<Card> = newDeck()

    /** Set by [deal] to the last card handed out, or by [startFromKnownHand]. */
    var trump: Card? = null
        private set

    val trumpSuit: Suit get() = requireNotNull(trump) { "the trump card is not known yet" }.suit

    /** Index of the trick being played, 0..9. */
    var currentRound: Int = 0
        private set

    /** The suit that has to be followed, null while the trick is still being led. */
    var roundSuit: Suit? = null
        private set

    private val _cardsPlayedInRound: MutableList<Card> = mutableListOf()
    val cardsPlayedInRound: List<Card> get() = _cardsPlayedInRound

    private val _tricks: MutableList<Trick> = mutableListOf()
    val tricks: List<Trick> get() = _tricks

    init {
        val built = mutableListOf<Player>()
        for (team in teams) {
            team.id.playerNames.forEachIndexed { offset, name ->
                val player = Player(
                    id = team.id.firstPlayerId + offset,
                    name = name,
                    team = team,
                    strategy = team.agent.strategy,
                    selfBeliefPolicy = config.selfBeliefPolicy,
                )
                player.isHuman = name == humanPlayerName
                team.addPlayer(player)
                built.add(player)
            }
        }
        players = built.sortedBy { it.id }

        // Randomise the seating, alternating between the teams, exactly like Game.__init__.
        for (team in teams) team.players.shuffle(random)
        val first = teams[random.nextInt(teams.size)]
        val second = if (first === teams[0]) teams[1] else teams[0]
        playersOrder = first.players.zip(second.players).flatMap { (a, b) -> listOf(a, b) }
    }

    fun playerNamed(name: String): Player? = players.firstOrNull { it.name == name }

    fun playerWithId(id: Int): Player = players[id - 1]

    // -----------------------------------------------------------------------------------------
    // Setting up
    // -----------------------------------------------------------------------------------------

    /**
     * Distributes the cards between the players. The last card dealt is the trump.
     *
     * Ported from `Game.hand_cards`, including the detail that the trump card simply stays in
     * the hand of whoever was dealt it last.
     */
    fun deal() {
        var last: Card? = null
        for (player in playersOrder) {
            repeat(CARDS_PER_HAND) {
                val card = deck.removeAt(random.nextInt(deck.size))
                player.addCard(card)
                player.team.initialPoints += card.points
                player.beliefs.observeDealt(card)
                last = card
            }
        }
        trump = last
        if (config.revealTrump) revealTrumpHolder(playersOrder.last())
    }

    /**
     * Sets the table up from the outside: one known hand, the trump card, and who sits where.
     *
     * This is how the advisor tracks a real game. The three other hands stay unknown
     * ([Player.handKnown] false) and the agents reason about them purely from beliefs.
     */
    fun startFromKnownHand(
        seat: Player,
        hand: List<Card>,
        trumpCard: Card,
        trumpHolder: Player?,
        leader: Player,
    ) {
        require(hand.size <= CARDS_PER_HAND) { "a Sueca hand holds at most $CARDS_PER_HAND cards" }

        for (player in players) {
            player.hand.clear()
            player.handKnown = player === seat
        }

        for (card in hand) {
            seat.addCard(card)
            seat.team.initialPoints += card.points
            for (player in players) player.beliefs.observeDealt(card, seat.id)
        }

        trump = trumpCard
        if (trumpHolder != null) revealTrumpHolder(trumpHolder)

        playersOrder = rotateToLeader(players.sortedBy { seatOrderIndex(it) }, leader)
        currentRound = 0
        roundSuit = null
        _cardsPlayedInRound.clear()
        _tricks.clear()
    }

    /**
     * Deterministic setup with all four hands known.
     *
     * Used to replay a game logged by the Python simulator, and to put the engine in front of a
     * specific position.
     */
    fun startFromHands(hands: Map<Int, List<Card>>, trumpCard: Card, leaderId: Int) {
        require(hands.keys.toSet() == (1..PLAYER_COUNT).toSet()) { "every seat needs a hand" }
        require(hands.values.flatten().toSet().size == hands.values.sumOf { it.size }) { "duplicate cards" }

        for (team in teams) {
            team.score = 0
            team.initialPoints = 0
        }

        for (player in players) {
            player.hand.clear()
            player.handKnown = true
            val hand = hands.getValue(player.id)
            for (card in hand) {
                player.addCard(card)
                player.team.initialPoints += card.points
                player.beliefs.observeDealt(card)
            }
        }

        trump = trumpCard
        if (config.revealTrump) {
            players.firstOrNull { trumpCard in it.hand }?.let { revealTrumpHolder(it) }
        }

        playersOrder = rotateToLeader(players.sortedBy { seatOrderIndex(it) }, playerWithId(leaderId))
        currentRound = 0
        roundSuit = null
        _cardsPlayedInRound.clear()
        _tricks.clear()
    }

    /** Playing order around the table: me, opponent, partner, opponent. */
    private fun seatOrderIndex(player: Player): Int = when (player.id) {
        1 -> 0      // Sporting first player
        3 -> 1      // Benfica first player
        2 -> 2      // Sporting second player
        else -> 3   // Benfica second player
    }

    private fun rotateToLeader(order: List<Player>, leader: Player): List<Player> {
        val index = order.indexOf(leader)
        require(index >= 0) { "$leader is not sitting at this table" }
        return order.drop(index) + order.take(index)
    }

    private fun revealTrumpHolder(holder: Player) {
        val card = trump ?: return
        for (player in players) player.beliefs.observeDealt(card, holder.id)
    }

    // -----------------------------------------------------------------------------------------
    // Playing
    // -----------------------------------------------------------------------------------------

    val currentPosition: Int get() = _cardsPlayedInRound.size

    val currentPlayer: Player get() = playersOrder[currentPosition]

    val isFinished: Boolean get() = currentRound >= ROUNDS_PER_GAME

    /** Everything the strategy of the player to act is allowed to see. */
    val turn: Turn get() = Turn(currentPosition, roundSuit, _cardsPlayedInRound.toList(), this)

    /** The cards the player to act is allowed to play. */
    fun legalCards(): List<Card> = currentPlayer.legalCards(roundSuit)

    /** What the current player's own strategy would do, without playing it. */
    fun suggestion(): Decision = currentPlayer.decide(turn)

    /**
     * Plays [card] for whoever is to act.
     *
     * Returns the finished trick when this was the fourth card, null otherwise.
     */
    fun play(card: Card): Trick? {
        check(!isFinished) { "the game is over" }
        val player = currentPlayer

        // A known hand that has run out is treated like an unknown one: it happens when a
        // session was started from fewer than ten cards, and refusing every card from then on
        // would strand the game rather than degrade it.
        if (player.handKnown && player.hand.isNotEmpty()) {
            require(card in legalCards()) { "${card.label} is not a legal card for ${player.name}" }
            player.removeCard(card)
        }

        if (_cardsPlayedInRound.isEmpty()) roundSuit = card.suit
        _cardsPlayedInRound.add(card)

        // Everyone updates their beliefs; the table each player owns decides what it does with it.
        val suit = roundSuit!!
        for (other in players) other.beliefs.observePlayed(card, suit, player.id)

        if (_cardsPlayedInRound.size < PLAYER_COUNT) return null

        val result = evaluateRound(_cardsPlayedInRound)
        val winner = playersOrder[result.winner]
        winner.team.score += result.points

        val trick = Trick(
            number = currentRound,
            plays = playersOrder.mapIndexed { index, seat -> Play(seat, _cardsPlayedInRound[index]) },
            winner = winner,
            points = result.points,
        )
        _tricks.add(trick)

        // Rotate the players order to the winner of the round.
        playersOrder = rotateToLeader(playersOrder, winner)
        _cardsPlayedInRound.clear()
        roundSuit = null
        currentRound++

        return trick
    }

    /** Lets the current player's agent choose, then plays it. */
    fun playAgentTurn(): Pair<Decision, Trick?> {
        val decision = suggestion()
        return decision to play(decision.card)
    }

    fun playToCompletion() {
        while (!isFinished) playAgentTurn()
    }

    /**
     * Calculates the points on the table and which of the cards is winning them.
     *
     * Also used by the strategies to score rounds that are still incomplete, and hypothetical
     * ones, which is why it accepts a partial list. The returned [RoundResult.winner] indexes
     * into [cards], which lines up with [playersOrder] because the cards played so far always
     * belong to the first N players in order.
     */
    fun evaluateRound(cards: List<Card>): RoundResult {
        require(cards.isNotEmpty()) { "cannot evaluate an empty round" }
        var winner = 0
        for (i in 1 until cards.size) {
            if (cards[i].beats(cards[winner], trumpSuit)) winner = i
        }
        return RoundResult(cards.sumOf { it.points }, cards[winner], winner)
    }

    fun result(): GameResult {
        val scores = teams.associate { it.id to it.score }
        val dealt = teams.associate { it.id to it.initialPoints }
        val winner = when {
            sporting.score > benfica.score -> TeamId.SPORTING
            benfica.score > sporting.score -> TeamId.BENFICA
            else -> null
        }
        return GameResult(winner, scores, dealt)
    }
}
